code/debugging/src/debug_logger.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+"""Dedicated debug logger node.
+
+It can receive messages via a multiprocessing ipc socket on localhost:52999.
+Message format is a python dict with keys {name, msg, level}.
+It will then send those messages into the ros console via rospy.log*()
+
+Main usecase: it enables not initialized python nodes to publish log messages
+to the /rosout topic.
+
+Without a fully initialized node,
+log messages from python files only show up in the ros logfile,
+but not in the /rosout topic and thus not in the rqt_console
+
+This is why this node is used by the debug_wrapper for logging
+before it has initialized its own node
+"""
+
+import time
+import sys
+from multiprocessing.connection import Listener, Client
+import threading
+import signal
+
+import rospy
+
+ADDRESS = ("localhost", 52999)
+AUTHKEY = b"debug_logger"
+CLOSE_MSG = "close"
+LISTENER_THREAD = None
+NODE_RUNNING = True
+
+MESSAGES_MUTEX = threading.Lock()
+MESSAGES = []
+
+
+def eprint(msg: str):
+    """Log msg into stderr.
+
+    Used instead of print, because only stderr seems to reliably land in agent.log
+
+    Args:
+        msg (str): Log message
+    """
+    print(f"[debug_logger]: {msg}", file=sys.stderr)
+
+
+def log(name: str, msg: str, level: str):
+    """Log to the ros console
+
+    Args:
+        name (str): Node name this message should be associated with
+        msg (str): Log message
+        level (str): Log level. One of (debug), info, warn, error, fatal.
+            debug level not recommended, because these are not published to /rosout
+    """
+    msg = f"[debug_logger for {name}]: {msg}"
+
+    level = level.lower()
+    if level == "debug":
+        rospy.logdebug(msg)
+    elif level == "info":
+        rospy.loginfo(msg)
+    elif level == "warn":
+        rospy.logwarn(msg)
+    elif level == "error":
+        rospy.logerr(msg)
+    else:
+        rospy.logfatal(msg)
+
+
+def run_listener(listener: Listener):
+    """Run the multiprocessing listener
+
+    The listener exits when it receives CLOSE_MSG
+
+    Args:
+        listener (Listener): Listener to run
+    """
+    running = True
+    while running:
+        try:
+            conn = listener.accept()
+            print(f"[debug_logger]: connection accepted from {listener.last_accepted}")
+            msg = None
+            if conn.poll(timeout=2.0):
+                msg = conn.recv()
+                if isinstance(msg, str):
+                    if msg.lower() == CLOSE_MSG:
+                        running = False
+                        msg = None
+            conn.close()
+        except Exception as error:
```

Comment:
What happens if the connection is not closed?

Reply:
@Zelberor do you think this is a real issue, or is this not relevant for our limited debugging use case.

Reply:
I don't know if it is really relevant, because an exception should only occur if the debug_wrapper fails while sending a log message which I deem unlikely.

But ignoring that, the [multiprocessing documentation](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.connection.Connection.close) tells me that the close() functions is automatically called when conn is garbage collected.

Now due to pythons weird scoping I think conn is not garbage collected before the listener tries to accept another connection. But the listener should be able to accept multiple connections (linux socket) and thus the accept fn will not fail due to conn still being active. At that point conn is reassigned and the old conn is garbage collected closing the connection.

In case the loops exists for some reason the garbage collection occurs when exiting the function.

So should be fine in all cases (speculation though, I don't know python that well)